src/services/documentRegistry.ts

```diff
@@ -194,14 +207,21 @@ namespace ts {
         function acquireOrUpdateDocument(
             fileName: string,
             path: Path,
-            compilationSettings: CompilerOptions,
+            compilationSettingsOrHost: CompilerOptions | MinimalResolutionCacheHost,
             key: DocumentRegistryBucketKey,
             scriptSnapshot: IScriptSnapshot,
             version: string,
             acquiring: boolean,
             scriptKind?: ScriptKind): SourceFile {
             scriptKind = ensureScriptKind(fileName, scriptKind);
+            const compilationSettings = getCompilationSettings(compilationSettingsOrHost);
+            const host: MinimalResolutionCacheHost | undefined = compilationSettingsOrHost === compilationSettings ? undefined : compilationSettingsOrHost as MinimalResolutionCacheHost;
             const scriptTarget = scriptKind === ScriptKind.JSON ? ScriptTarget.JSON : getEmitScriptTarget(compilationSettings);
+            const sourceFileOptions: CreateSourceFileOptions = {
+                languageVersion: scriptTarget,
+                impliedNodeFormat: host && getImpliedNodeFormatForFile(path, host.getCompilerHost?.()?.getModuleResolutionCache?.()?.getPackageJsonInfoCache(), host, compilationSettings),

```

Comment:
Have you verified that all these optionals are actually set in built-in scenarios where we care about perf, like TS Server and `--watch` (uh, is this used for watch? I have no idea)? I was surprised to find a case recently where the module resolution cache was _not_ being passed through where it easily could have been, though I think I already fixed that in a recent PR. Point being, it’s worth checking, because it would be pretty bad if every call to this function in TS Server had to go searching the file system for package.jsons because the cache didn’t make it through.

Reply:
Hm, a valid `host` is provided everywhere, that much I've checked, however I have not audited the actual presence of these optionals, no. Easiest way would be to make them not-optional and let the compiler do the work, I guess. AFAIK, the only concrete implementation in the language service is a `ConfiguredProject`, so that's the authoritative location to check, I suppose.

Reply:
A `ConfiguredProject` always has a `CompilerHost` set by services, and such a host is 1. always freshly made, and 2. forwards the `LanguageServiceHost`'s `getModuleResolutionCache`, which means it's the `ConfigredProject`'s cache, which always exists.